.circleci/config.yml

```diff
@@ -1,5 +1,171 @@
-version: 2
+version: 2.1
+
+executors:
+  go:
+    working_directory: ~/go/src/github.com/hashicorp/nomad
+    docker:
+      - image: dantoml/nomad-build-image:e8d69b1e8
+  go-machine:
+    working_directory: ~/go/src/github.com/hashicorp/nomad
+    machine:
+      image: circleci/classic:201808-01
+  docker-builder:
+    working_directory: ~/go/src/github.com/hashicorp/nomad
+    machine: true # TODO: Find latest docker image id
+
 jobs:
+  build-deps-image:
+    executor: docker-builder
+    steps:
+      - checkout
+      - run: docker pull dantoml/nomad-build-image:latest || true
+      - run: docker build -t dantoml/nomad-build-image:$CIRCLE_SHA1 ./.circleci/build-images/go
+      - run: docker push dantoml/nomad-build-image:$CIRCLE_SHA1
+
+  lint:
+    executor: go
+    environment:
+      GOPATH: /home/circleci/go
+    steps:
+      - checkout
+      - run: make check
+
+  test-container:
+    executor: go
+    parameters:
+      test_packages:
+        type: string
+        default: ""
+      exclude_packages:
+        type: string
+        default: ""
+    environment:
+      GOTEST_PKGS: "<< parameters.test_packages >>"
+      GOTEST_PKGS_EXCLUDE: "<< parameters.exclude_packages >>"
+      GOPATH: /home/circleci/go
+      GOTESTSUM_JUNITFILE: /tmp/results.xml
+      GOMAXPROCS: 1
+      NOMAD_SLOW_TEST: 1
+    steps:
+      - checkout
+      - run: |
+          if [ -z $GOTEST_PKGS_EXCLUDE ];
+          then
+            unset GOTEST_PKGS_EXCLUDE
+          else
+            unset GOTEST_PKGS
+          fi
+          make test-nomad
+      - store_test_results:
+          path: /tmp
+
+  test-rkt:
+    executor: go-machine
+    environment:
+      GOTEST_PKGS: "./drivers/rkt"
+      GOTESTSUM_JUNITFILE: /tmp/results.xml
+      GOPATH: /home/circleci/go
+      GOVERSION: 1.12.1
+      PROTOC_VERSION: 3.6.1
+      GOMAXPROCS: 1
+      NOMAD_SLOW_TEST: 1
+      RKT_VERSION: 1.29.0
+    steps:
+      - checkout
+      - run:
+          name: install go
+          command: |
+            sudo rm -rf /usr/local/go
+            wget https://dl.google.com/go/go$GOVERSION.linux-amd64.tar.gz
+            sudo tar -C /usr/local -xzf go$GOVERSION.linux-amd64.tar.gz
+      - run:
+          name: install protoc
+          command: |
+            sudo rm -rf /usr/bin/protoc
```

Comment:
much of the installation scripts here (e.g. go, protoc, rkt, consul, vault) are also done in ./scripts.  I would suggest reusing rather than inline them in yaml.

Reply:
I semi-deliberately ignored the travis scripts because a lot of them are fairly travis specific, and they're already not shared between vagrant and travis, and this lets us more easily have a matrix of dependency versions fairly quickly if wanted.

Reply:
I feel like there is a lot of duplication here and and not all scripts are travis specific.  For example, there are 3 installers for protoc in this PR and three places to set its version, in addition to the `./scripts/vagrant-linux-priv-protoc.sh` that's already being used in our release docker image docker file, and I don't believe we intend to do matrix dependency testing for protoc binary.

Additionally, scripts inside yaml are harder to check with linter tools.

I'd suggest refactoring scripts to minimize current travis/vagrant specificity and reuse rather than have ~4 almost identical snippets laying around.

Reply:
I just dry'd it up and moved protoc invocation to the script. If they're going to be used in more than one place, they should be unnamespaced though.